sentinel.go

```diff
@@ -289,50 +301,93 @@ func (sc *Sentinel) ensureClients(conn Conn) error {
 		return err
 	}
 
-	newSecsM := map[string]bool{}
+	newSecsM := map[string]Client{}
 	for _, secM := range secMM {
 		newSecAddr, err := sentinelMtoAddr(secM, "SENTINEL SLAVES")
 		if err != nil {
 			return err
 		}
-		newSecsM[newSecAddr] = true
+		newSecsM[newSecAddr] = nil
 	}
 
 	return sc.setClients(newPrimAddr, newSecsM)
 }
 
-// one day we might keep track of the secondary clients in here, not sure yet
-func (sc *Sentinel) setClients(newPrimAddr string, newSecsM map[string]bool) error {
+func (sc *Sentinel) setClients(newPrimAddr string, newSecsM map[string]Client) error {
+
+	var stateChanged bool
+	var toClose []Client
+	var newPrim Client
+
 	sc.l.RLock()
 	oldPrimAddr := sc.primAddr
-	secsAdded, secsRemoved := sc.secsDiff(sc.secsM, newSecsM)
-	sc.l.RUnlock()
+	oldPrim := sc.prim
+
+	// for each actual Client instance in sc.secsM, either move it over to
+	// newSecsM (if the address is shared) or make sure it is closed if it's
+	// not becoming the primary.
+	for addr, client := range sc.secsM {
+		if client == nil || addr == newPrimAddr {
+			// do nothing
+		} else if _, ok := newSecsM[addr]; ok {
+			newSecsM[addr] = client
+		} else {
+			toClose = append(toClose, client)
+			stateChanged = true
+		}
+	}
 
-	primChanged := oldPrimAddr != newPrimAddr
-	if !primChanged && len(secsAdded) == 0 && len(secsRemoved) == 0 {
-		return nil
+	// this is only checks if a secondary was added so we know the replica set
+	// state has changed later in the method.
+	for addr := range newSecsM {
+		if _, ok := sc.secsM[addr]; !ok {
+			stateChanged = true
+		}
+	}
+
+	// if the primary is now a secondary it can be moved into newSecsM too,
+	// otherwise it can be closed. oldPrim might be nil if this is the first
+	// initialization.
+	if _, ok := newSecsM[oldPrimAddr]; ok {
+		newSecsM[oldPrimAddr] = oldPrim
+	} else if oldPrim != nil {
+		toClose = append(toClose, oldPrim)
+	}
 
-	} else if !primChanged {
-		sc.l.Lock()
-		sc.secsM = newSecsM
-		sc.l.Unlock()
+	// if the primary has then record that the replica set state has changed,
+	// and while we're here try to pre-emtively populate newPrim from the old
+	// secondary set.
+	if oldPrimAddr != newPrimAddr {
```

Comment:
There is quite a bit of code just because the primary is kept separate from all the other clients. Instead we could keep them all in one map and then just lookup the primary by `sc.prim` when needed. This way we could treat all clients the same (we would need to skip the primary in `Addrs`, but that's just a simple `if addr == sc.primAddr`)

What do you think?

Reply:
went ahead and did that, you're right it did simplify the code quite a bit